### ISSUE TITLE
Column configuration

### ISSUE THREAD

**Issue by pbugni** [uwcirg/cosri-patientsearch#143]
We need a flexible configuration system to define which columns and the FHIR source of each.

A few examples:
MRN - a column that doesn't exist in COSRI, but others will need, and presumably pull from a Patient.identifier with a particular `system` value.
Alerts - a source FHIR type that will bring patients to the top of the list with important messages for end users.

**Comment by mcjustin** [comment]
DCW will utilize this via https://www.pivotaltracker.com/story/show/183147834

**Comment by mcjustin** [comment]
@ivan-c  to add straw horse config which includes the default fields.

**Comment by ivan-c** [comment]
@achen2401 @pbugni @mcjustin @lober how does something like this look? Happy to make edits, but tried to represent "hardcoded" columns (no `expr`), alongside "configurable" ones:

```
DASHBOARD_COLUMNS=
[
    {
        "label": "First Name"
    },
    {
        "label": "Last Name"
    },
    {
        "label": "Gender",
        "expr": "$.Patient.gender"
    },
    {
        "label": "Birth Date"
    },
    {
        "label": "Last Accessed"
    },
    {
        "label": "MRN", 
        "expr": "$.Patient.identifier[?(@.system='https://hl7.org/mrn')].value"
    }
]
```

**Comment by achen2401** [comment]
@ivan-c I am wondering if the syntax for the FHIR path in your example is referencing [this](http://hl7.org/fhirpath/)?   Do you know if [this](https://github.com/hl7/fhirpath.js/) is the library that people are using to evaluate FHIR path?

**Comment by ivan-c** [comment]
@achen2401 sorry I kept editing my comment, the expressions I included in my example were JSONPath, not FHIRPath. There are a lot more libraries that implement JSONPath than FHIRPath. However, I took a quick look at the FHIRPath libs you linked, and didn't see any reason they shouldn't work.

**Comment by mcjustin** [comment]
@ivan-c  weren't we aiming for the labels to be configurable strings, independent of the reference to the field? It looks like your labels are acting as field references, at least for the "hardcoded" columns (no `expr`).

**Comment by ivan-c** [comment]
> weren't we aiming for the labels to be configurable strings, independent of the reference to the field? It looks like your labels are acting as field references, at least for the "hardcoded" columns (no `expr`).

@mcjustin Good point- we wouldn't be able to rename the labels for the hard-coded columns with the example I gave. Maybe we could use some hard-coded constants, one for each hard-coded field, eg:

```
    {
        "label": "Birth Date"
        "expr": "BIRTH_DATE"
    },

```

**Comment by lober** [comment]
This seems like a difficult way to have a discussion.  I'm willing to do it, but it seems like GitHub issues might be best used **first** to identify and issue then to point to where it is being discussed, related materials, etc. and finally to the documentation of the agreed upon approach to resolve the issue.  Not everything is that complex, of course, but I can't figure out what the current example of DASHBOARD_COLUMNS is, what FHIRPath loos like, etc.   I'd like to be able to propose some additional column attributes.


On a separate note (also not a good place for questions...?!) do FHIRPath expressions always evaluate to the contents of an item in a resource, or can they also evaluate to True or False?

**Comment by achen2401** [comment]
@lober I tried out the config structure proposed by Ivan, with the expressions or `expr` written in [JSON Path](https://goessner.net/articles/JsonPath/), this is an example using the current fields in the FEMR patient list (and tested and working in my instance):
```
DASHBOARD_COLUMNS = [{
      'label': 'First Name',
      'expr': '$.name[0].given[0]',
    },
    {
      'label': 'Last Name',
      'expr': '$.name[0].family',
    },
    {
      'label': 'Birth Date',
      'expr': '$.birthDate',
    },
    {
      'label': 'Last Accessed',
      'defaultSort': 'desc',
      'expr': '$.meta.lastUpdated',
      'dataType': 'date',
    }]
```
I added two additional attributes, i.e.  `defaultSort` and `dataType` (the names can be changed). The `defaultSort` attribute tells the code to sort the patient list in FEMR by the Last Accessed field in descending order (which is what it currently is being sorted) and a `dataType` attribute of date allows the code to recognize the field as date and reformats the output text accordingly.   
And to answer your question 

> do FHIRPath expressions always evaluate to the contents of an item in a resource, or can they also evaluate to True or False?

Currently the expressions evaluate to the contents of an item in a resource, and I _don't think_ they can be evaluated to be true or false, but I could be wrong.

**Comment by mcjustin** [comment]
> do FHIRPath expressions always evaluate to the contents of an item in a resource, or can they also evaluate to True or False?

I do see some boolean operations at https://hl7.org/fhir/fhirpath.html , eg "hasValue", "conformsTo", "memberOf".

**Comment by mcjustin** [comment]
Noting that we didn't decide between these two:
- FHIRPath - note that CQL is a superset of FHIRPath, so using this may educate our CQL usage.
- JSONPath - more tooling available, bigger community.

**Comment by mcjustin** [comment]
@sierramatice  You can see a couple configurations of this at https://femr.isacc.dev.cirg.uw.edu and https://femr.helloworld.dev.cirg.uw.edu (test/test).

**Comment by sierramatice** [comment]
@mcjustin I added a mock up and comments to the PT story

**Comment by mcjustin** [comment]
@pbugni  mock-up from @sierramatice  is at https://www.pivotaltracker.com/story/show/183147834/comments/233809643
